zarr/v3/codecs/__init__.py

```diff
@@ -86,72 +79,87 @@ def _validate_codecs(codecs: List[Codec], array_metadata: CoreArrayMetadata) ->
                     f"ArrayArrayCodec '{type(codec)}' cannot follow after "
                     + f"BytesBytesCodec '{type(prev_codec)}'."
                 )
-
-            if isinstance(codec, ShardingCodec):
-                assert len(codec.configuration.chunk_shape) == len(array_metadata.shape), (
-                    "The shard's `chunk_shape` and array's `shape` need to have the "
-                    + "same number of dimensions."
-                )
-                assert all(
-                    s % c == 0
-                    for s, c in zip(
-                        array_metadata.chunk_shape,
-                        codec.configuration.chunk_shape,
-                    )
-                ), (
-                    "The array's `chunk_shape` needs to be divisible by the "
-                    + "shard's inner `chunk_shape`."
-                )
             prev_codec = codec
 
-        if any(isinstance(codec, ShardingCodec) for codec in codecs) and len(codecs) > 1:
+        if any(isinstance(codec, ShardingCodec) for codec in self.codecs) and len(self.codecs) > 1:
             warn(
                 "Combining a `sharding_indexed` codec disables partial reads and "
                 + "writes, which may lead to inefficient performance."
             )
 
-    def _array_array_codecs(self) -> List[ArrayArrayCodec]:
-        return [codec for codec in self.codecs if isinstance(codec, ArrayArrayCodec)]
-
-    def _array_bytes_codec(self) -> ArrayBytesCodec:
-        return next(codec for codec in self.codecs if isinstance(codec, ArrayBytesCodec))
-
-    def _bytes_bytes_codecs(self) -> List[BytesBytesCodec]:
-        return [codec for codec in self.codecs if isinstance(codec, BytesBytesCodec)]
-
-    async def decode(self, chunk_bytes: BytesLike) -> np.ndarray:
-        for bb_codec in self._bytes_bytes_codecs()[::-1]:
-            chunk_bytes = await bb_codec.decode(chunk_bytes)
+    def _codecs_with_resolved_metadata(
+        self, chunk_metadata: ArraySpec
+    ) -> Iterator[Tuple[Codec, ArraySpec]]:
+        for codec in self.codecs:
+            yield (codec, chunk_metadata)
+            chunk_metadata = codec.resolve_metadata(chunk_metadata)
+
+    async def decode(
+        self,
+        chunk_bytes: BytesLike,
+        chunk_metadata: ArraySpec,
+        runtime_configuration: RuntimeConfiguration,
+    ) -> np.ndarray:
+        codecs = list(self._codecs_with_resolved_metadata(chunk_metadata))[::-1]
```

Comment:
do you think it would be easier to reason about this if we split the codec pipeline into three groups:

1. `BytesBytesCodec`s
2. `BytesArrayCodec`s
3. `ArrayArrayCodec`s

That would cut down on the number of instance checks later on.

Reply:
alternatively, could we do a tagged union by adding a `codec_type` attribute with type `Literal["BytesBytes", "BytesArray", "ArrayArray"]`? then we just need 1 class definition and attribute checks instead of `isinstance`

Reply:
ah, my proposal is actually not an alternative to @jhamman's idea. i think we could do both of these things.

Reply:
Worth noting that the proposal to partition the codecs by type actually brings us back to the design of hdf5 / v2, where there is a separate filter pipeline (filters basically being `ArrayArray` transformations).

Reply:
Not sure it makes anything more readable: https://github.com/zarr-developers/zarr-python/pull/1632/commits/b825b4fab6043e3b36b13b9638aa58ade7c20057

Reply:
An attribute check would be better if we wanted to enable developers to use codecs without requiring that they inherit from our base classes (we could also achieve this with structural subtyping). And `isinstance(x, `ArrayArrayCodec`) is teeny bit more work to read / write than `x.codec_type == "ArrayArray"`. That being said, I don't think either of these complaints should block your efforts here, so feel free to ignore